File: utilities/dates/java/src/main/java/io/polyglotutilities/dates/DateUtils.java
package io.polyglotutilities.dates;

import java.time.Duration;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeParseException;

public final class DateUtils {
    private DateUtils() {
    }

    public static OffsetDateTime parseDate(String value) {
        if (value == null || value.trim().isEmpty()) {
            return null;
        }
        try {
            return OffsetDateTime.parse(value.trim());
        } catch (DateTimeParseException exception) {
            try {
                return OffsetDateTime.of(
                        java.time.LocalDateTime.parse(value.trim()),
                        ZoneOffset.UTC
                );
            } catch (DateTimeParseException ignored) {
                return null;
            }
        }
    }

    public static Long daysBetween(String start, String end) {
        OffsetDateTime startDate = parseDate(start);
        OffsetDateTime endDate = parseDate(end);
        if (startDate == null || endDate == null) {
            return null;
        }
        return Math.round(Duration.between(startDate, endDate).toSeconds() / 86400.0);
    }
}

File: utilities/dates/kotlin/src/main/kotlin/io/polyglotutilities/dates/DateUtils.kt
package io.polyglotutilities.dates

import java.time.Duration
import java.time.LocalDateTime
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.time.format.DateTimeParseException
import kotlin.math.floor

object DateUtils {
    fun parseDate(value: String?): OffsetDateTime? {
        if (value.isNullOrBlank()) {
            return null
        }

        val normalized = value.trim()
        return try {
            OffsetDateTime.parse(normalized)
        } catch (_: DateTimeParseException) {
            try {
                LocalDateTime.parse(normalized).atOffset(ZoneOffset.UTC)
            } catch (_: DateTimeParseException) {
                null
            }
        }
    }

    fun daysBetween(start: String?, end: String?): Long? {
        val startDate = parseDate(start)
        val endDate = parseDate(end)
        if (startDate == null || endDate == null) {
            return null
        }

        val days = Duration.between(startDate, endDate).seconds / 86400.0
        return floor(days + 0.5).toLong()
    }
}
